server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -546,18 +553,38 @@ public Builder addIndexShard(IndexShardRoutingTable indexShard) {
          */
         public Builder addShard(ShardRouting shard) {
             assert shard.index().equals(index) : "cannot add [" + shard + "] to routing table for " + index;
-            IndexShardRoutingTable indexShard = shards.get(shard.id());
+            int shardId = shard.id();
+            ensureShardArray(shardId);
+            IndexShardRoutingTable.Builder indexShard = shards[shardId];
             if (indexShard == null) {
-                indexShard = new IndexShardRoutingTable.Builder(shard.shardId()).addShard(shard).build();
+                shards[shardId] = IndexShardRoutingTable.builder(shard.shardId()).addShard(shard);
             } else {
-                indexShard = new IndexShardRoutingTable.Builder(indexShard).addShard(shard).build();
+                indexShard.addShard(shard);
             }
-            shards.put(indexShard.shardId().id(), indexShard);
             return this;
         }
 
+        void ensureShardArray(int shardId) {
+            if (shards == null) {
+                shards = new IndexShardRoutingTable.Builder[shardId + 1];
+            } else if (shards.length < shardId + 1) {
+                IndexShardRoutingTable.Builder[] updated = new IndexShardRoutingTable.Builder[shardId + 1];
+                System.arraycopy(shards, 0, updated, 0, shards.length);
+                shards = updated;
+            }
+        }
+
         public IndexRoutingTable build() {
-            return new IndexRoutingTable(index, shards);
+            final IndexShardRoutingTable[] res;
+            if (shards != null) {
+                res = new IndexShardRoutingTable[shards.length];
+                for (int i = 0; i < shards.length; i++) {
+                    res[i] = shards[i].build();
+                }
+            } else {
+                res = new IndexShardRoutingTable[0];
```

Comment:
I'm not sure, I don't really see how it would be. But annoyingly we have a couple of tests that trigger this with an empty builder randomly. I put down a todo to look into cleaning this up. Maybe we can become a little cleverer here still anyway with how we handle the array :)